File: uuResources.r
# \file      uuResources.r
# \brief     Resources and tiers.
# \author    Harm de Raaff
# \copyright Copyright (c) 2017, Utrecht University. All rights reserved.
# \license   GPLv3, see LICENSE.

# Resources and tiers
#
# A tier designates a cost level of storage, i.e. resource.
# Some storage is more expensive than others.

# A tier is added as metadata to a resources
# A resources can only have 1 tier.


# \brief uuResourceExistst - check whether given resource actually exists
#
# \param[in] *resourceName
# \param[out] *exists
#
uuResourceExists(*resourceName, *exists)
{
        *exists = false;

        foreach(*row in SELECT RESC_ID, RESC_NAME WHERE RESC_NAME = '*resourceName') {
                *exists = true;
		succeed;
        }
}

# \brief uuSetResourceTier
#
# \param[in] 	*resourceName
# \param[in] 	*tierName
# \param[out] 	*result
# \param[out] 	*errorInfo
#
uuSetResourceTier(*resourceName, *tierName, *result, *errorInfo)
{
	*result = 'ok';

	# 1)First check whether resource actually exists
        uuResourceExists(*resourceName, *rescExists)

        if (!*rescExists) {
                *result = '-1'; # Resource does not exist.
		*errorInfo = 'Resource *resourceName does not exist';
                succeed;
        }


        # 2)Check whether tier- metadata exists for given resource based upon 'org_storage_tier' as meta attribute
        *metaFound = false;
	*metaName = UURESOURCETIERATTRNAME;
        foreach(*row in SELECT RESC_ID, RESC_NAME, META_RESC_ATTR_NAME, META_RESC_ATTR_VALUE WHERE RESC_NAME='*resourceName' AND META_RESC_ATTR_NAME='*metaName' ) {
                *metaFound = true;
        }

	msiString2KeyValPair("", *kvpResc);
        msiAddKeyVal(*kvpResc, *metaName, *tierName);

        if (!*metaFound ) {
                *err = msiAssociateKeyValuePairsToObj( *kvpResc, *resourceName, "-R");

		if (*err!=0 ) {
			*result='-999';
			*errorInfo = 'Something went wrong adding tier metadata';
			succeed;
		}
        }
        else {
		*err = msiSetKeyValuePairsToObj( *kvpResc, *resourceName, "-R");

		if (*err!=0 ) {
                        *result='-999';
                        *errorInfo = 'Something went wrong updating tier metadata';
                        succeed;
                }
        }
}

# \brief List of  all resources and their tier/storage data (if present).
#        Tiers are only assigned to the resources that are allowed.
#        Therefore, no further restriction has to be added for the type of resource (which should be storage)
#
uuListResourceTiers(*result, *errorInfo)
{
        *result = 0;

	*foundStandardTier = false;
        *allRescTiers = list();

        # fetch tier information for all resources and filter duplicates
	*metaName = UURESOURCETIERATTRNAME;
        foreach(*row in SELECT META_RESC_ATTR_VALUE WHERE  META_RESC_ATTR_NAME = '*metaName' ) {
        	# writeLine('stdout', *row.META_RESC_ATTR_VALUE);
                *allRescTiers = cons(*row.META_RESC_ATTR_VALUE, *allRescTiers);
		if (*row.META_RESC_ATTR_VALUE == 'Standard') {
			*foundStandardTier = true;
		}
        }

	if (!*foundStandardTier) { # Add standard tier if not found in database
		*allRescTiers = cons('Standard', *allRescTiers);
	}

        *allRescTiers;
}


# \brief List of all existing resources TODO: exclude coordination resources.
#
uuListResources()
{
        *allResources = list();

        foreach(*row in SELECT RESC_ID, RESC_NAME ) {
                msiString2KeyValPair("", *kvp);
                *kvp.resourceId = *row.RESC_ID;
                *kvp.resourceName = *row.RESC_NAME;
                *allResources = cons(*kvp, *allResources);
        }

        *allResources;
}


# \brief
#
# \param[in] *kvpList - list with all key-value pairs to be deleted
# \param[in] *objectName - description as known in dbs
# \param[in] *objectType - description as known within iRODS {-u, -C, etc}
#
uuRemoveKeyValuePairList(*kvpList, *objectName, *objectType, *status, *statusInfo)
{
        *status = 'Success';
        *statusInfo = '';

	foreach (*kvp in *kvpList) {
                 *err = errormsg( msiRemoveKeyValuePairsFromObj(*kvp, *objectName, *objecType), *errmsg);
                 if (*err < 0) {
                         *status = 'ErrorDeletingMonthlyStorage';
                         *statusInfo = 'Error deleting metadata: *err - *errmsg';
                         succeed;
                 }
	}
}


# \brief For all categories known store all found storage data for each group belonging to those category.
#        Store as metadata on group level holding
#        1) category of group on probe date - this can change
#        2) tier
#        3) actual calculated storage for the group
#
uuStoreMonthlyStorageStatistics(*status, *statusInfo)
{
	writeLine('serverLog', 'Start uuStoreMonthlyStorageStatistics');

	# Really for the frontend but can be of use for this as well
	*status = 'Success';
	*statusInfo = '';

	*month = uuGetCurrentStatisticsMonth();

	*metadataName = UUMETADATASTORAGEMONTH ++ *month;  #UUORGMETADATAPREFIX ++ 'storageDataMonth' ++ *month;

	# First delete all previous data for this month-number
	*kvpList = list();
	foreach (*row in SELECT META_USER_ATTR_VALUE, USER_GROUP_NAME WHERE META_USER_ATTR_NAME = '*metadataName' ) {

                msiString2KeyValPair("", *kvp);
                msiAddKeyVal(*kvp, *metadataName, *row.META_USER_ATTR_VALUE);
		*kvpList = cons(*kvp, *kvpList);
                 *err = errormsg( msiRemoveKeyValuePairsFromObj(*kvp, *row.USER_GROUP_NAME, "-u"), *errmsg);
                 if (*err < 0) {
	                 *status = 'ErrorDeletingMonthlyStorage';
        	         *statusInfo = 'Error deleting metadata: *err - *errmsg';
                         succeed;
                 }
	}
	# Problem here is that *objectName differs for vkp's.
	# So it should be taken along in building the kvp list
	#uuRemoveKeyValuePairList(*kvpList, *objectName, *objectType, *status, *statusInfo);


        # zone is used to search in proper paths for storage
        *zone =  $rodsZoneClient;

        #uuListCategories(*listCategories)

	*listCategories = uuListCategories();

        # Get all existing tiers for initialisation purposes per category
        *allTiers = uuListResourceTiers(*result, *errorInfo);

        *kvpResourceTier = uuKvpResourceAndTiers();

         #per group find the storage amount for
         # 1) dynamic storage and
         # 2) vault
         *storageCalculationSteps = list("dynamic", "vault");

	# 3) Furthermore each group's revision store has to be taken into account

	# Step through all categories
        foreach (*categoryName in *listCategories) {
                *listGroups = list();
                uuListGroupsOnCategory(*categoryName, *listGroups);

                #per group find the storage amount for
                # 1) dynamic storage and
                # 2) vault
                foreach(*groupName in *listGroups ) {
                        msiString2KeyValPair("", *groupTierStorage);
                        foreach (*tier in *allTiers){ # initialize group tier storage
                                *groupTierStorage."*tier" = '0';
                        }

                        foreach (*step in *storageCalculationSteps) {
                                if (*step == 'dynamic') {
                                        *collName = '/*zone/home/'++ *groupName;
                                }
                                else { # vault sitation - strip groupname down to its basic name (research-)
                                        uuGetGroupNameForVault(*groupName, *groupNameVault);
                                        *collName = '/*zone/home/vault'++ *groupNameVault;
                                }

			        # Per group two statements are required to gather all data
         			# 1) data in folder itself
         			# 2) data in all subfolders of the folder
				# 3) data from revisions of that group

				# 1) Collect all data in folder itself
                               	foreach (*row in SELECT SUM(DATA_SIZE), RESC_NAME WHERE COLL_NAME = '*collName') {
                                       	# This brings the total for dynamic storage of a group per RESOURCE

                                    	*thisResc = *row.RESC_NAME;
                                        *thisTier = *kvpResourceTier."*thisResc";

                                        # Totals on group level
                                        *newGroupSize = double(*groupTierStorage."*thisTier") + double(*row.DATA_SIZE);
                                        *groupTierStorage."*thisTier" = str(*newGroupSize);
				}

				# 2) Collect all data in all subfolders of the folder
                                foreach (*row in SELECT SUM(DATA_SIZE), RESC_NAME WHERE COLL_NAME like '*collName/%%') {
                                        # This brings the total for dynamic storage of a group per RESOURCE

                                        *thisResc = *row.RESC_NAME;
                                        *thisTier = *kvpResourceTier."*thisResc";

                                        # Totals on group level
                                        *newGroupSize = double(*groupTierStorage."*thisTier") + double(*row.DATA_SIZE);
                                        *groupTierStorage."*thisTier" = str(*newGroupSize);
                                }
			}

			# 3) Collect all data in revision folder of this group
			# This can be caught in a single statement as the group folder itself does not hold data
                        *revisionCollName = '/*zone' ++ UUREVISIONCOLLECTION ++ '/' ++ *groupName ++ '/%%';
                        foreach (*row in SELECT SUM(DATA_SIZE), RESC_NAME WHERE COLL_NAME like '*revisionCollName') {
                                # This brings the total for dynamic storage of a group per RESOURCE
                                *thisResc = *row.RESC_NAME;
                                *thisTier = *kvpResourceTier."*thisResc";

                                 # Totals on group level
                                 *newGroupSize = double(*groupTierStorage."*thisTier") + double(*row.DATA_SIZE);
                                 *groupTierStorage."*thisTier" = str(*newGroupSize);
                        }

                        # Group information complete.
			# Add it to dbs
                        foreach (*tier in *allTiers) {
                                msiString2KeyValPair("", *kvpGroupStorage);
                                *storage = *groupTierStorage."*tier";
                                *json_str = '["*categoryName","*tier","*storage"]';
                                *kvpGroupStorage."*metadataName" = *json_str;
                                *err = errormsg(msiAssociateKeyValuePairsToObj( *kvpGroupStorage, *groupName, "-u"), *errmsg);
				if (*err < 0) {
					*status = 'ErrorWritingMonthlyStorage';
					*statusInfo = 'Error adding metadata: *err - *errmsg';
					succeed;
				}
                        }
		}
	}
}


# \brief Returns the number of the month {'01',...'12'} that currently is the month reporting is about.
#
uuGetCurrentStatisticsMonth()
{
        msiGetIcatTime(*timestamp, "icat");
        *month = int(timestrf(datetime(int(*timestamp)), "%m"));

	# Format month as '01'-'12'
	*strMonth = str(*month);
	if (strlen(*strMonth)==1) {
		*strMonth = '0' ++ *strMonth;
	}
	*strMonth;
}


# \brief Returns *kvp with resourceName as key and tierName as value.
#        This way it is easy to use the name of a resource as an index and retrieve the corresponding tierName.
uuKvpResourceAndTiers()
{
	*listResources = uuListResources();

	msiString2KeyValPair("", *kvp);

	*metaName = UURESOURCETIERATTRNAME;
	foreach (*resource in *listResources) {

	# Because outerjoins are impossible in iRods and there is nog guarantee that all resources have org_m

		*resourceName = *resource.resourceName;
		*kvp."*resourceName" = 'Standard';

		*sqlResource = *resource.resourceName;
                foreach(*row in SELECT RESC_ID, RESC_NAME, META_RESC_ATTR_NAME, META_RESC_ATTR_VALUE WHERE RESC_NAME='*sqlResource' AND META_RESC_ATTR_NAME = '*metaName' ) {
                        *kvp."*resourceName" = *row.META_RESC_ATTR_VALUE;
                }
	}

	*kvp;
}


# \brief Get a list of all known categories where current user is datamanager of.
#        Returns list of categories for this user as a datamanager.
#
uuListCategoriesDatamanager()
{
        *listCategories = list();
        *user = uuClientFullName;

        # Get categories current user is a datamanager of
        foreach (
                *row in
                SELECT USER_NAME
# \param[out] *status           -return status to frontend
# \param[out] *statusInfo       -return specific information regarding *status
                WHERE  USER_TYPE            = 'rodsgroup'
                        AND USER_NAME like 'datamanager-%%'
        ) {
                *datamanagerGroupName = *row.USER_NAME;
                uuGroupUserExists(*datamanagerGroupName, *user, true, *membership)
		if (*membership) {
			# datamanagerGroupName must be deciphered to get actual category (its prefixed with 'datamanager-'
			*partsCategories = split(*datamanagerGroupName,'-');
			*category = '';
			*counter = 0;
			foreach (*cat in *partsCategories) {
				if (*counter > 0) {
					if (*counter==1) {
						*category = *cat;
					}
					else {
						*category = *category ++ '-' ++ *cat;
					}
				}
				*counter = *counter + 1;
			}
			*listCategories = cons(*category, *listCategories);
		}
        }

	*listCategories;
}


# \brief Get a list of all known categories
#
uuListCategories()
{
        *listCategories = list();
        foreach (*row in SELECT META_USER_ATTR_VALUE
                WHERE  USER_TYPE            = 'rodsgroup'
                  AND  META_USER_ATTR_NAME  = 'category') {

		*listCategories = cons(*row.META_USER_ATTR_VALUE, *listCategories);
        }
	*listCategories;
}


# \brief List of groups.
#
# \param[in]  *categoryName
# \param[out] *listGroups
#
uuListGroupsOnCategory(*categoryName, *listGroups)
{
	*listGroups = list();
       foreach (
                *row in
                SELECT USER_NAME
                WHERE  USER_TYPE            = 'rodsgroup'
                  AND  META_USER_ATTR_NAME  = 'category'
                  AND  META_USER_ATTR_VALUE = '*categoryName'
        ) {

		*listGroups = cons(*row.USER_NAME, *listGroups);
        }
}

# \brief Get the base group name, stripped off of 'research-' etc.
#
# \param[in]  *groupName - full name of a group including 'research-' etc
# \param[out] *groupBase
#
uuGetGroupNameForVault(*groupName, *groupBase)
{
	*partsList = split(*groupName,'-');

         *count = 0;
         *groupBase = '';
         foreach (*part in *partsList) {
         	if (*count>0) {
                        *groupBase = *groupBase ++ '-' ++ *part;
                }
                *count = *count + 1;
         }
}

File: uuResources.py
# \file      uuResources.py
# \brief     Functions for statistics module - in essence a python extension directly related to uuResources.r
# \author    Lazlo Westerhof
# \author    Felix Croes
# \author    Harm de Raaff
# \copyright Copyright (c) 2018-2019 Utrecht University. All rights reserved.
# \license   GPLv3, see LICENSE.

import os
from datetime import datetime
import json

import genquery
import irods_types


def uuRuleGetResourceTierData(rule_args, callback, rei):
    """Get JSON represenation of resource and its tier info."""
    resourceName = rule_args[0]

    tierName = getTierOnResourceName(resourceName, callback)

    rule_args[1] = json.dumps({"resourceName": resourceName,
                               "org_storage_tier": tierName})


def uuRuleGetResourcesAndTierData(rule_args, callback, rei):
    """Get all resources and their tier data as a json representation."""
    resourceList = list()

    iter = genquery.row_iterator(
        "RESC_ID, RESC_NAME",
        "",
        genquery.AS_LIST, callback
    )

    for row in iter:
        resourceId = row[0]
        resourceName = row[1]
        tierName = getTierOnResourceName(resourceName, callback)
        resourceList.append({'resourceName': resourceName,
                             'resourceId': resourceId,
                             'org_storage_tier': tierName})

    rule_args[0] = json.dumps(resourceList)


def uuRuleGetMonthStoragePerTierForGroup(rule_args, callback, rei):
    """Get json representation for storage data for a period of 12 months for a specific group.

       Storage is per month and tier
       Format is "month=12-tier=Standard": "222222222222"
    """
    groupName = rule_args[0]
    currentMonth = int(rule_args[1])  # this is the month that came from the frontend

    allStorage = []  # list of all month-tier combinations present including their storage size

    # per month gather month/tier/storage information from metadata:
    # metadata-attr-name = UUMETADATASTORAGEMONTH + '01'...'12'
    # metadata-attr-val = [category,tier,storage] ... only tier and storage required wihtin this code
    for counter in range(0, 11):
        referenceMonth = currentMonth - counter
        if referenceMonth < 1:
            referenceMonth = referenceMonth + 12

        metadataAttrNameRefMonth = UUMETADATASTORAGEMONTH + '%0*d' % (2, referenceMonth)

        iter = genquery.row_iterator(
            "META_USER_ATTR_VALUE, USER_NAME, USER_GROUP_NAME",
            "META_USER_ATTR_NAME = '" + metadataAttrNameRefMonth + "' AND USER_NAME = '" + groupName + "'",
            genquery.AS_LIST, callback
        )

        for row in iter:
            data = json.loads(row[0])

            tierName = data[1]
            data_size = data[2]  # no construction for summation required in this case

            key = 'month=' + str(referenceMonth) + '-tier=' + tierName
            allStorage.append({key: data_size})

    rule_args[2] = json.dumps(allStorage)


def uuRuleGetMonthlyStorageStatistics(rule_args, callback, rei):
    """Collect storage data for all categories."""
    categories = getCategories(callback)

    rule_args[0] = getMonthlyCategoryStorageStatistics(categories, callback)


def uuRuleGetMonthlyStorageStatisticsDatamanager(rule_args, callback, rei):
    """Collect storage data for a datamanager."""
    datamanagerUser = rule_args[0]
    categories = getCategoriesDatamanager(datamanagerUser, callback)

    rule_args[1] = getMonthlyCategoryStorageStatistics(categories, callback)


def uuRuleGetAllGroupsForDatamanager(rule_args, callback, rei):
    """Get all groups for all categories a person is datamanager of."""
    datamanagerUser = rule_args[0]
    categories = getCategoriesDatamanager(datamanagerUser, callback)

    datamanagerGroups = getGroupsOnCategories(categories, callback)

    rule_args[1] = json.dumps(datamanagerGroups)


def uuRuleExportMonthlyCategoryStatisticsDM(rule_args, callback, rei):
    """Collect storage stats for all twelve months based upon categories a user is datamanager of:
       - Category
       - Subcategory
       - Groupname
       - Tier
       - 12 columns, one per month, with used storage count in bytes
    """
    datamanagerUser = rule_args[0]
    categories = getCategoriesDatamanager(datamanagerUser, callback)
    allStorage = []

    # Select a full year by not limiting UUMETADATASTORAGEMONTH to a perticular month. But only on its presence.
    # There always is a maximum of one year of history of storage data
    for category in categories:
        groupToSubcategory = {}

        iter = genquery.row_iterator(
            "META_USER_ATTR_VALUE, META_USER_ATTR_NAME, USER_NAME, USER_GROUP_NAME",
            "META_USER_ATTR_VALUE like '[\"" + category + "\",%' AND META_USER_ATTR_NAME like  '" + UUMETADATASTORAGEMONTH + "%'",
            genquery.AS_LIST, callback
        )

        for row in iter:
            attrValue = row[0]
            month = row[1]
            month = str(int(month[-2:]))  # the month storage data is about, is taken from the attr_name of the AVU
            groupName = row[3]

            # Determine subcategory on groupName
            try:
                subcategory = groupToSubcategory[groupName]
            except KeyError:
                catInfo = groupGetCategoryInfo(groupName, callback)
                subcategory = catInfo['subcategory']
                groupToSubcategory[groupName] = subcategory

            temp = json.loads(attrValue)
            category = temp[0]
            tier = temp[1]
            storage = int(temp[2])

            allStorage.append({'category': category,
                               'subcategory': subcategory,
                               'groupname': groupName,
                               'tier': tier,
                               'month': month,
                               'storage': str(storage)})

    rule_args[1] = json.dumps(allStorage)


def groupGetCategoryInfo(groupName, callback):
    """Get category and subcategory for a group.

       Returns a dict with indices 'category' and 'subcategory'.
    """
    category = ''
    subcategory = ''

    iter = genquery.row_iterator(
        "META_USER_ATTR_NAME, META_USER_ATTR_VALUE",
        "USER_GROUP_NAME = '" + groupName + "' AND  META_USER_ATTR_NAME LIKE '%category'",
        genquery.AS_LIST, callback
    )

    for row in iter:
        attrName = row[0]
        attrValue = row[1]

        if attrName == 'category':
            category = attrValue
        elif attrName == 'subcategory':
            subcategory = attrValue

    return {'category': category, 'subcategory': subcategory}


def getMonthlyCategoryStorageStatistics(categories, callback):
    """Collect storage stats of last month only.

       Storage is summed up for each category/tier combination.
       JSON presentation: Array ( [0] => Array ( [category] => initial [tier] => Standard [storage] => 15777136 )
    """
    month = '%0*d' % (2, datetime.now().month)
    metadataName = UUMETADATASTORAGEMONTH + month
    storageDict = {}

    for category in categories:
        iter = genquery.row_iterator(
            "META_USER_ATTR_VALUE, META_USER_ATTR_NAME, USER_NAME, USER_GROUP_NAME",
            "META_USER_ATTR_VALUE like '[\"" + category + "\",%' AND META_USER_ATTR_NAME = '" + metadataName + "'",
            genquery.AS_LIST, callback
        )

        for row in iter:
            # hier wordt door alle groepen gezocht, geordend van een category.
            # per tier moet worden gesommeerd om totale hoeveelheid storage op een tier te verkrijgen.
            attrValue = row[0]

            temp = json.loads(attrValue)
            category = temp[0]
            tier = temp[1]
            storage = int(float(temp[2]))

            try:
                storageDict[category][tier] = storageDict[category][tier] + storage
            except KeyError:
                callback.writeString('serverLog', 'Exception')
                # if key error, can be either category or category/tier combination is missing
                try:
                    storageDict[category][tier] = storage
                except KeyError:
                    storageDict[category] = {tier: storage}

    # prepare for json output, convert storageDict into dict with keys
    allStorage = []

    for category in storageDict:
        for tier in storageDict[category]:
            allStorage.append({'category': category,
                               'tier': tier,
                               'storage': str(storageDict[category][tier])})

    return json.dumps(allStorage)


def getGroupsOnCategories(categories, callback):
    """Get all groups belonging to all given categories."""
    groups = []
    metadataAttrNameRefMonth = UUMETADATASTORAGEMONTH + '%0*d' % (2, datetime.now().month)

    for category in categories:
        iter = genquery.row_iterator(
            "USER_NAME",
            "USER_TYPE = 'rodsgroup' AND META_USER_ATTR_NAME = 'category' AND META_USER_ATTR_VALUE = '" + category + "' ",
            genquery.AS_LIST, callback
        )

        for row in iter:
            groupName = row[0]

            iter2 = genquery.row_iterator(
                "META_USER_ATTR_VALUE, USER_NAME, USER_GROUP_NAME",
                "META_USER_ATTR_NAME = '" + metadataAttrNameRefMonth + "' AND USER_NAME = '" + groupName + "'",
                genquery.AS_LIST, callback
            )

            data_size = 0
            for row in iter2:
                data = row[0]
                temp = json.loads(data)
                data_size = data_size + int(temp[2])  # no construction for summation required in this case

            groups.append([groupName, data_size])

    return groups


def getCategoriesDatamanager(datamanagerName, callback):
    """Get all categories for curent datamanager."""
    categories = []

    iter = genquery.row_iterator(
        "USER_NAME",
        "USER_TYPE = 'rodsgroup' AND USER_NAME like 'datamanager-%'",
        genquery.AS_LIST, callback
    )

    for row in iter:
        # @TODO membership still has to be checked
        datamanagerGroupname = row[0]
        temp = datamanagerGroupname.split('-')  # 'datamanager-initial' is groupname of datamanager, second part is category
        categories.append(temp[1])

    return categories


# \brief
#
def getCategories(callback):
    """Get all categories currently present."""
    categories = []

    iter = genquery.row_iterator(
        "META_USER_ATTR_VALUE",
        "USER_TYPE = 'rodsgroup' AND  META_USER_ATTR_NAME  = 'category'",
        genquery.AS_LIST, callback
    )

    for row in iter:
        categories.append(row[0])

    return categories


def getTierOnResourceName(resourceName, callback):
    """Get Tiername, if present, for given resource.

       If not present, fall back to default tier name.
    """
    tierName = UUDEFAULTRESOURCETIER  # Add default tier as this might not be present in database.

    # find (possibly present) tier for this resource
    iter = genquery.row_iterator(
        "RESC_ID, RESC_NAME, META_RESC_ATTR_NAME, META_RESC_ATTR_VALUE",
        "RESC_NAME = '" + resourceName + "' AND META_RESC_ATTR_NAME = '" + UURESOURCETIERATTRNAME + "'",
        genquery.AS_LIST, callback
    )

    for row in iter:
        tierName = row[3]

    return tierName
